kibana/idvametrics/analytics_script.py

```diff
@@ -0,0 +1,285 @@
+from datetime import datetime, timedelta
+from dateutil import parser
+from opensearchpy import OpenSearch, helpers
+
+ANALYTICS_INDEX_PATTERN = "dev-analytics"
+SK_INDEX_PATTERN = "dev-skevents-*"
+DEFAULT_NUM_COMPOSITE_BUCKETS = 10
+FIVE_MINS = timedelta(minutes=5)
+ONE_DAY = timedelta(days=1)
+
+query_avg_response_time = {
+    "query": {
+        "bool": {
+            "must": [{"match_phrase": {"flowId": {}}}, {"range": {"tsEms": {}}}],
+            "filter": [{"match_all": {}}],
+        }
+    },
+    "_source": ["companyId", "flowId"],
+    "size": 1,
+    "aggs": {
+        "composite_buckets": {
+            "aggs": {
+                "min": {"min": {"field": "tsEms"}},
+                "max": {"max": {"field": "tsEms"}},
+                "sessionLength": {
+                    "bucket_script": {
+                        "buckets_path": {"startTime": "min", "endTime": "max"},
+                        "script": "params.endTime - params.startTime",
+                    }
+                },
+            },
+            "composite": {
+                "sources": [{"agg": {"terms": {"field": "interactionId.keyword"}}}]
+            },
+        },
+    },
+}
+
+
+def update_nested_key(dictionary: dict, key_path: list, value: dict):
+    """
+    Updates a dictionary at the nested key defined by key_path with value, adding
+    the nested keys if they do not exist
+    """
+    curr_dict = dictionary
+    for key in key_path:
+        curr_dict = curr_dict.setdefault(key, {})
+
+    curr_dict.update(value)
+
+
+def get_most_recent_timestamp(flow_id: str):
+    """
+    We want to retrieve all data from dev-skevents with a timestamp at any time or more recent
+    than 5 minutes before the timestamp of the most recent data in dev-analytics-*. This
+    ensures that we do not miss any data.
+    """
+
+    newest_timestamp_query = {
+        "query": {
+            "bool": {
+                "must": [
+                    {
+                        "match_phrase": {"flowId": flow_id},
+                    },
+                ],
+            },
+        },
+        "size": 1,
+        "sort": [
+            {
+                "tsEms": {"order": "desc"},
+            },
+        ],
+    }
+
+    res = elasticsearch.search(
+        index=f"{ANALYTICS_INDEX_PATTERN}-*", body=newest_timestamp_query
+    )
+    try:
+        # Attempting to get the time of five minutes prior to the most recent timestamp (tsEms)
+        # of a document in the analytics index.
+        most_recent_timestamp = parser.parse(res["hits"]["hits"][0]["_source"]["tsEms"])
+        five_mins_prior = most_recent_timestamp - FIVE_MINS
+        return int(five_mins_prior.timestamp())
+    except KeyError:
+        # There is nothing in the analytics index for the given flow_id, so we want the timestamp
+        # of five minutes prior to the current time.
+        current_timestamp = (datetime.now() - FIVE_MINS).timestamp()
+        return int(current_timestamp)
```

Comment:
Only line 89 could trigger a `KeyError`, the other lines need to be moved out of the try/except block. As a general rule of thumb, don't include code in try/catch blocks that's not related to the possible exception you're trying to handle.
```suggestion
    try:
        # Attempting to get the time of five minutes prior to the most recent timestamp (tsEms)
        # of a document in the analytics index.
        latest_timestamp = parser.parse(res["hits"]["hits"][0]["_source"]["tsEms"])
    except KeyError:
        # Nothing in the analytics index for the given flow_id, so use the current time.
        latest_timestamp = datetime.now()
        
    five_mins_prior = latest_timestamp - FIVE_MINS
    return int(five_mins_prior.timestamp())
```